Sharing relay

knowuh opened the pull request:
This is a big change to look at.

The summary is basically this: use `new SharingRelay(null, app);`  to create a sharing container.

`SharingClient` s create a new iFramePhone if one isn't passed in as a first arguement.
`SharinRelay` is a subclass of SharingClient that sets up SharingConnections to all of its child iFrames

PublishResponses now include children.

knowuh committed:
Working on a mediated middle for sharing relay / proxy.

knowuh committed:
fixed email address in npm info

knowuh committed:
Introduce `SharingRelay`  probably the only class one needs to use actually.

dougmartin commented on src/demo/demo.html:
```diff
@@ -4,6 +4,14 @@
+        console.log("dom node insterted");
+        console.log(evt);
+      }
+      document.addEventListener("DOMSubtreeModified", listener, false);
```
Did you mean to keep this debug call here?

dougmartin commented on src/child-connection.ts:
```diff
@@ -0,0 +1,43 @@
+  context: Context;
+  id: number;
+
+  constructor(
```
I wonder if we should change the parameters to the constructor a single hash of options instead of a parameter list.  I think it makes the client code more self-documenting.

dougmartin commented on src/child-connection.ts:
```diff
@@ -0,0 +1,43 @@
+    this.phone.disconnect();
+  }
+  sendPublish() {
+    if(this.connection) {
```
How can connection be falsey here?  It is created unconditionally in the constructor.

dougmartin commented on src/init-message.ts:
```diff
@@ -1,10 +1,11 @@
-
+export type ISODateString    = string;
 
+export const InitMessageName = "SharinatorInit";
```
It would help me in the codap-shim app if the messages kept a common prefix *and* that prefix was exported from init-message.  I have a nearly blind proxy that pumps everything but the Sharinator messages.  Something like this:

```
export const MessagePrefix = "Sharinator"
export const InitMessageName = `${MessagePrefix}Init`
export const InitResponseMessageName = `${MessagePrefix}InitResponse`
```

dougmartin commented on src/sharing-client.ts:
```diff
@@ -1,43 +1,80 @@
 
-  constructor(phone:IFramePhone, app:SharableApp) {
-    this.phone = phone;
+  constructor(phone:IFramePhoneUp|null, app:SharableApp) {
```
Why not reverse the parameters and make phone optional like `phone?:IFramePhoneUp`?

dougmartin commented on src/sharing-parent.ts:
```diff
@@ -1,24 +1,42 @@
+
 
+export type PublishResultsCallback = (p:PublishResponse) => void;
+export type InitResponseCallback = (initRepy: InitResponseMessage) => void;
```
I think that is supposed to be `initReply`

dougmartin commented on src/sharing-relay.ts:
```diff
@@ -0,0 +1,91 @@
+        };
+      });
+    });
+    Promise.all(promises).then( () => this.sendPublishResponse(this.childResponses));
```
This one line is what makes promises awesome  👍

dougmartin commented on src/sharing-relay.ts:
```diff
@@ -0,0 +1,91 @@
+    _.each(this.calledChildren, (c) => {
+      c.disconnect();
+    });
+    this.calledChildren = [];
```
I think you need to add `this.connectedChildren = []` here too

dougmartin reviewed (changes requested):
Lots of comments :)

knowuh commented:
Thanks for these change requests Doug, very helpful.

knowuh committed:
Use object params / typed params for constructors.

* ChildConnection, SharingParent, SharingClient, and SharingRelay all use typed params in constructor.
* Phone is now optional param for SharingClient
* Other minor fixes.

knowuh committed:
Add `publicationListeners`.  Use SharingRelay as top Parent. Fix recursion & listening bugs.

- PublicationListeners are interested in receiving Publications. (The top level container might add one to its SharingRelay)

- We probably shouldn't use `SharingParent` or `SharingClient` any more.  Instead always use `SharingRelay`.

knowuh committed:
Disconnect only connected children.  Reset calledChildren.

knowuh committed:
Move publishable and init-messages into simple `types` file.  Stop importing from `index`.

Also remove console logging.

knowuh committed:
Demo: Remove unused pieces. Nicer styling for larger iframes.

knowuh committed:
Remove demo source.

knowuh committed:
Use `MessagePrefix` for post message API message names.

All messages now start with "Sharinator"
